### PR TITLE
HOD-151 | Added IF Connector

### PR DESCRIPTION
**Pull request by owen-opencast**
Moved IfConnectorSpec to component:tests
Added IfConnector, domain models and validation
Updated spec tests
Resolved deprecated Logger warnings
Updated handling of 404s - 404s with a NO_DATA_FOUND msg in body are passed on - income and employments returned empty list in this event, but these endpoints don't return lists. Standard 404s are swallowed - indicates something wrong with our URLs.

### CODE REVIEW

File: app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+package uk.gov.hmrc.organisationsdetailsapi.domain.integrationframework
+
+import play.api.libs.functional.syntax._
+import play.api.libs.json.Reads._
+import play.api.libs.json.{Format, JsPath}
+
+
+case class AccountingPeriod(apStartDate: String, apEndDate: String, turnover: Int)
+
+case class CreateCorporationTaxReturnDetailsRequest(utr: String, taxpayerStartDate: String, taxSolvencyStatus: String, accountingPeriods: Seq[AccountingPeriod])
```
Comment (parkeradam) [inline app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala:26]: create? Unsure what this is being used for.

File: app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        (JsPath \ "turnover").read[Int]
+      )(AccountingPeriod.apply _),
+    (
+      (JsPath \ "apStartDate").write[String] and
+        (JsPath \ "apEndDate").write[String] and
+        (JsPath \ "turnover").write[Int]
+      )(unlift(AccountingPeriod.unapply))
+  )
+
+  implicit val createCorporationTaxReturnDetailsRequestFormat = Format[CreateCorporationTaxReturnDetailsRequest](
```
Comment (parkeradam) [inline app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala:50]: Format for questioned case class.

File: app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/EmployeeCountResponse.scala [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+case class EmployeeCountRequest(startDate: String, endDate: String, references: Seq[PayeReference])
+
+case class PayeReferenceAndCount(districtNumber: String, payeReference: String, counts: Seq[Count])
+
+case class PayeReference(districtNumber: String, payeReference: String)
+
+
+/*
+employeeCount:
```
Comment (ajchilts) [inline app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/EmployeeCountResponse.scala:39]: Do we need this comment?

File: app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/EmployeeCountResponse.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+case class EmployeeCountResponse(startDate: String, endDate: String, references: Seq[PayeReferenceAndCount])
+
+case class EmployeeCountRequest(startDate: String, endDate: String, references: Seq[PayeReference])
+
+case class PayeReferenceAndCount(districtNumber: String, payeReference: String, counts: Seq[Count])
+
+case class PayeReference(districtNumber: String, payeReference: String)
+
+
+/*
```
Comment (parkeradam) [inline app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/EmployeeCountResponse.scala:38]: Please remove comment. Think the object below shows the validation for this.

**Review by ajchilts — APPROVED** [review]

File: test/component/uk/gov/hmrc/organisationsdetailsapi/connectors/IfConnectorSpec.scala [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+  val saReturn = createValidSelfAssessmentReturnDetails()
+  val employeeCountRequest = createValidEmployeeCountRequest()
+  val employeeCountResponse = createValidEmployeeCountResponse()
+
+  val invalidTaxReturn  = createValidCorporationTaxReturnDetails().copy(utr = "")
+  val invalidSaReturn = createValidSelfAssessmentReturnDetails().copy(utr = "")
+  val invalidEmployeeCountRequest = createValidEmployeeCountRequest().copy(startDate = "")
+  val invalidEmployeeCountResponse = createValidEmployeeCountResponse().copy(startDate = "")
+
+  "fetch foo" should {
```
Comment (parkeradam) [inline test/component/uk/gov/hmrc/organisationsdetailsapi/connectors/IfConnectorSpec.scala:116]: test name should be more specific.

**Review by parkeradam — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/EmployeeCountResponse.scala (hunk 3 above)
Comment (owen-opencast) [reply app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/EmployeeCountResponse.scala:39]: Good spot

File: app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala (hunk 1 above)
Comment (owen-opencast) [reply app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala:26]: Removed - snuck over from IF stub

File: app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala (hunk 2 above)
Comment (owen-opencast) [reply app/uk/gov/hmrc/organisationsdetailsapi/domain/integrationframework/CorporationTaxReturnDetails.scala]: As above

File: test/component/uk/gov/hmrc/organisationsdetailsapi/connectors/IfConnectorSpec.scala (hunk 5 above)
Comment (owen-opencast) [reply test/component/uk/gov/hmrc/organisationsdetailsapi/connectors/IfConnectorSpec.scala]: Updated

**Review by parkeradam — APPROVED** [review]

**Pull request merged**